### ISSUE TITLE
sql: crash on nil ctx when cleaning up the flow infrastructure

### ISSUE THREAD

**Issue by jimethn** [cockroachdb/cockroach#58303]
**Describe the problem**

During a heavy query the node crashed with a sql panic. 

**To Reproduce**

The query was a count(distinct) on one table with an inner merge join against another table, with various where clauses, against a large data set.

**Expected behavior**
If we hit max-sql-memory, I'd expect cockroachdb to kill the query, rather than lose the whole node.

**Additional data / screenshots**

```
E201224 19:49:53.378265 86683757 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 174080 bytes requested, 16
095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:53.395598 86683705 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 409600 bytes requested, 16
095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:53.463929 86683727 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 174080 bytes requested, 16
095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:53.482758 86683707 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 409600 bytes requested, 16
095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:53.898687 86683919 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 174080 bytes requested, 16
095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:53.984845 86683770 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 174080 bytes requested, 16
095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:54.000274 86683944 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 409600 bytes requested, 16095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:54.096445 86683986 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 174080 bytes requested, 16095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:54.182005 86683975 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 174080 bytes requested, 16095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:54.200636 86683884 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 409600 bytes requested, 16095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:54.291254 86683976 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 409600 bytes requested, 16095359596 currently allocated, 16095365120 bytes in budget
E201224 19:49:54.678548 86683866 sql/flowinfra/flow_registry.go:250 ⋮ [n15] flow id:‹79beb4bd-2386-4b6b-917f-b1eb3b7e620a› : 2 inbound streams timed out after 10s; propagated error throughout flow
E201224 19:49:54.895955 86683978 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 573440 bytes requested, 16095195756 currently allocated, 16095365120 bytes in budget
E201224 19:49:55.864321 86645974 sql/distsql/server.go:368 ⋮ [n15,client=‹10.42.12.0:40857›,hostssl,user=‹query_user›] error setting up flow: not enough memory to setup vectorized plan: ‹root›: memory budget exceeded: 327680 bytes requested, 16095165036 currently allocated, 16095365120 bytes in budget
E201224 19:49:55.864945 86684309 sql/distsql/server.go:368 ⋮ [n15] error setting up flow: ‹root›: memory budget exceeded: 573440 bytes requested, 16095226476 currently allocated, 16095365120 bytes in budget
E201224 19:49:55.865912 86645974 sql/conn_executor.go:806 ⋮ [n15,client=‹10.42.12.0:40857›,hostssl,user=‹query_user›] a SQL panic has occurred while executing the following statement:
‹SELECT count(DISTINCT (fp.id)) AS count_by_id FROM sometable AS fp INNER MERGE JOIN othertable AS a ON ((fp.c = a.c)
 AND (fp.id = a.id)) AND (a.d IS NULL) WHERE (((fp.c = $1) AND (fp.l >= $2::TIMESTAMP)) AND ((
((fp.state NOT IN ($3,)) AND (fp.r NOT IN ($4,))) AND (fp.m NOT IN ($5,))) AND (fp.p NOT IN ($6, $7, $8))))›
E201224 20:05:11.522483 62498897 util/log/crash_reporting.go:138 ⋮ [n11,client=‹10.42.11.0:42795›,hostssl,user=‹query_user›] a panic has occurred!
runtime error: invalid memory address or nil pointer dereference
E201224 19:49:55.866405 86645974 util/log/crash_reporting.go:138 ⋮ [n15,client=‹10.42.12.0:40857›,hostssl,user=‹us_2b_vm_query_user›] a panic has occurred!
runtime error: invalid memory address or nil pointer dereference
(1) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:506
  | [...repeated from below...]
Wraps: (2) while executing: SELECT count(DISTINCT (_._)) AS _ FROM _ AS _ INNER MERGE JOIN _ AS _ ON ((_._ = _._) AND (_._ = _._)) AND (_._ IS NULL) WHERE (((_._ = $1) AND (_._ >= $2::TIMESTAMP)) AND ((((_._ NOT IN ($3,)) AND (_._ NOT IN ($4,))) AND (_._ NOT IN ($5,))) AND (_._ NOT IN ($6, $7, __more1__))))
Wraps: (3) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:506
  | runtime.gopanic
  |     /usr/local/go/src/runtime/panic.go:679
  | runtime.panicmem
  |     /usr/local/go/src/runtime/panic.go:199
  | runtime.sigpanic
  |     /usr/local/go/src/runtime/signal_unix.go:394
  | github.com/opentracing/opentracing-go.SpanFromContext
  |     /go/src/github.com/cockroachdb/cockroach/vendor/github.com/opentracing/opentracing-go/gocontext.go:22
  | github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog
  |     /go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:93
  | github.com/cockroachdb/cockroach/pkg/util/log.vEventf
  |     /go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:243
  | github.com/cockroachdb/cockroach/pkg/util/log.VEventf
  |     /go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:271
  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).ReleaseLeases
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:1193
  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).ReleaseAll
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:1205
  | github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).NewFlowContext.func2
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:453
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).setupFlows.func1
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:175
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).setupFlows
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:267
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:397
  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1002
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1001
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:872
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:203
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1533
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1535
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1391
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:508
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  |     /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:626
  | runtime.goexit
  |     /usr/local/go/src/runtime/asm_amd64.s:1357
Wraps: (4) runtime error: invalid memory address or nil pointer dereference
Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.errorString
E201224 19:49:55.885495 86645974 util/log/crash_reporting.go:321 ⋮ [n15,client=‹10.42.12.0:40857›,hostssl,user=‹query_user›] Queued as error ‹5424a72413124da99a148af7cafce31b›
```

**Environment:**
 - CockroachDB version: v20.2.2
 - Server OS: Amazon Linux 2
 - Docker image: cockroachdb/cockroach:v20.2.2
 - Instance type: i3.2xlarge (8 CPU, 61GiB RAM)
 - Cluster size: 6 nodes
 - Client app: JDBC

**Additional context**
I get that we exceeded max-sql-memory, but the issue is that doing so caused the node to crash. There may be an unhandled error?

**Comment by yuzefovich** [comment]
Thanks for the report. Note that the crash isn't caused directly by the hitting `max-sql-memory` budget but by `nil` context object (it is possible that the former caused the latter). Looking into it.

**Comment by yuzefovich** [comment]
Hm, upon the code inspection I don't see why this crash happened, so I wonder whether @jimethn could reproduce it and share the reproduction steps with us. Without a repro we probably won't be able to make progress here.

**Comment by asubiotto** [comment]
Hi @jimethn,

When you can, could you share repro instructions?

**Comment by yuzefovich** [comment]
I believe this was fixed by #59003 which will be included in 20.2.5.
